### PR TITLE
Implement modular trainer stack and configuration restructure

### PR DESCRIPTION
**Pull request by mbaetiong**
This pull request introduces a modular, Hydra-configurable training stack for Codex ML, refactors configuration files for improved composability and backwards compatibility, and adds targeted documentation and test coverage for the new trainer stack. The changes also update Semgrep security rules for more precise Python pattern matching. The most important changes are grouped below.

**1. Modular Training Stack & Config Refactor**

* Added new modules: `src/modeling.py` (centralizes Hugging Face model/tokenizer loading with LoRA/PEFT support), `src/training/trainer.py` (mixed-precision aware trainer with evaluation, gradient accumulation, logging, and checkpointing), `src/logging_utils.py` (TensorBoard/MLflow integration), and `src/data/datasets.py` (TSV-based text classification DataLoader).
* Refactored Hydra config files (`configs/default.yaml`, `configs/model/base.yaml`, `configs/training/base.yaml`, `configs/data/tiny.yaml`) to use defaults lists for modular composition, while preserving legacy keys as aliases for backward compatibility [[1]](diffhunk://#diff-e69eb813cac10cbe6915b6ca52647f37ee3eb7e0a8821fcd76f23aa503a9b2f5L3-R32) [[2]](diffhunk://#diff-74c3fcbdd14c84a516a33f5256e6cd0b44ea8324b9627370b432b757d679ce18L1-R35) [[3]](diffhunk://#diff-6ed746302659a50ba133997c0a43720cb7c4bf9b4182acf35b87edda61e9e108L3-R9) [[4]](diffhunk://#diff-6ed746302659a50ba133997c0a43720cb7c4bf9b4182acf35b87edda61e9e108R22-R30) [[5]](diffhunk://#diff-6ed746302659a50ba133997c0a43720cb7c4bf9b4182acf35b87edda61e9e108R52-R68) [[6]](diffhunk://#diff-5476acaecb699541efc89e8957b1b6231f763b4c09a728ed1089694994bc91c4L1-R20).
* Updated documentation in `README.md`, `docs/modules/model_registry.md`, and `docs/reproducibility.md` to describe the new modular stack, configuration, and reproducibility workflow [[1]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R60-R88) [[2]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R181-R202) [[3]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R222-R243) [[4]](diffhunk://#diff-1fe4f5ddf315de6adf9684e7cbf28bfa3729742e219c5a99907161ada1bd31d9L22-R35) [[5]](diffhunk://#diff-457404b5a680ea73803e9c71ef32c6218c78554c70d99cd8b63430d2be95356fL36-R39) [[6]](diffhunk://#diff-457404b5a680ea73803e9c71ef32c6218c78554c70d99cd8b63430d2be95356fL50-R50) [[7]](diffhunk://#diff-457404b5a680ea73803e9c71ef32c6218c78554c70d99cd8b63430d2be95356fL63-R63) [[8]](diffhunk://#diff-457404b5a680ea73803e9c71ef32c6218c78554c70d99cd8b63430d2be95356fL83-R83) [[9]](diffhunk://#diff-457404b5a680ea73803e9c71ef32c6218c78554c70d99cd8b63430d2be95356fL119-R119) [[10]](diffhunk://#diff-457404b5a680ea73803e9c71ef32c6218c78554c70d99cd8b63430d2be95356fL173-R173).

**2. Testing and Logging Enhancements**

* Added focused trainer stack tests under `tests/` and a new `nox` session (`tests_trainer`) to run them, improving CI and local reproducibility [[1]](diffhunk://#diff-0af764ef14cf4e51bc49bad670bb745eec536c32ddd5ecab6ec0606258ac6092R1-R10) [[2]](diffhunk://#diff-f7a16a65f061822bcc73b8296f4dc837353d379d8d9cc5307982cb6941442835R214-R230).
* Introduced `error_log.md` for capturing model and API errors during training and evaluation.

**3. Semgrep Security Rule Improvements**

* Updated Python Semgrep rules to use explicit pattern matching for subprocess and requests usage, improving detection of insecure patterns [[1]](diffhunk://#diff-71e7b9c66d33418b7e1eb736efb9db083e87c840aab370092379f3341e6acc26L7-R13) [[2]](diffhunk://#diff-71e7b9c66d33418b7e1eb736efb9db083e87c840aab370092379f3341e6acc26L30-R43) [[3]](diffhunk://#diff-5b304dc8776b704e0d02282fd57e9c01a402a48db96b9c1f91147070bb97ef69L6-R14) [[4]](diffhunk://#diff-5b304dc8776b704e0d02282fd57e9c01a402a48db96b9c1f91147070bb97ef69R25-R26) [[5]](diffhunk://#diff-5bdffba6fcdaf0042adc2f0a931909ca866e7789ce52f9edfb964bb8c049af52L7-R21).

**4. Documentation and Change Logging**

* Added a detailed `change_log.md` summarizing the modular stack, config changes, and new tests.

These changes modernize the training pipeline, improve reproducibility, and enhance security and test coverage for Codex ML.

------
## Summary
- add modeling, logging, dataset, and trainer modules with PEFT-ready initialization, evaluation loops, and checkpoint retention
- restructure Hydra defaults into dedicated model/training/data configs while preserving legacy keys
- expand unit coverage for the new trainer stack and refresh documentation/error logging assets

## Testing
- `PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 pytest tests/modeling/test_modeling_module.py tests/logging/test_logging_utils_module.py tests/data/test_datasets_module.py tests/training/test_extended_trainer.py tests/security/test_semgrep_rules.py`
- `PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 pytest tests/data/test_datasets_module.py tests/training/test_extended_trainer.py`

------
https://chatgpt.com/codex/tasks/task_e_68effe90a6748331a969c1b9c4afb2d8

### CODE REVIEW

File: src/training/__init__.py [hunk 1]
```diff
@@ -1 +1,10 @@
 """Training utilities supporting offline audits."""
+
+from .trainer import CheckpointConfig, ExtendedTrainer, TrainerConfig
+
+__all__ = [
+    "CheckpointConfig",
+    "ExtendedTrainer",
+    "Trainer",
```
[the remark this answers is not in the document]
Comment (mbaetiong) [reply src/training/__init__.py]: @codex fix comment

File: src/modeling.py [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+class ModelInitConfig:
+    """High-level configuration for model + tokenizer initialisation."""
+
+    model_name: str
+    tokenizer_name: str | None = None
+    dtype: str = "float32"
+    device: str = "auto"
+    trust_remote_code: bool = False
+    load_config: Mapping[str, Any] = field(default_factory=dict)
+    lora: LoraSettings = field(default_factory=LoraSettings)
```
[the remark this answers is not in the document]
Comment (mbaetiong) [reply src/modeling.py]: @codex fix comment

File: src/logging_utils.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        except Exception as exc:  # pragma: no cover - offline guard
+            LOGGER.debug("Failed to end MLflow run cleanly: %s", exc)
+
+
+__all__ = [
+    "LoggingConfig",
+    "LoggingSession",
+    "log_metrics",
+    "setup_logging",
+    "shutdown_logging",
```
[the remark this answers is not in the document]
Comment (mbaetiong) [reply src/logging_utils.py]: @codex fix comment
